lang/query/src/lib.rs

```diff
@@ -6,62 +9,133 @@ use rust_lapper::Lapper;
 pub use result::Error;
 
 mod asserts;
+mod cache;
+mod edit;
+mod fs;
 mod info;
+mod lift;
+mod load;
+mod modules;
 mod result;
-mod view;
-mod view_mut;
+mod rt;
+mod spans;
+mod xfunc;
+
+use cache::*;
 
+pub use edit::*;
+pub use fs::*;
 pub use info::*;
-pub use view::*;
-pub use view_mut::*;
+pub use load::*;
+pub use xfunc::*;
 
-#[derive(Default)]
+/// A database tracking a set of source files
 pub struct Database {
-    files: HashMap<Url, codespan::File<String>>,
+    /// The source provider of the files (file system or in-memory)
+    source: Box<dyn FileSource>,
+    /// The source code text of each file
+    files: Cache<codespan::File<String>>,
+    /// The CST of each file (once parsed)
+    cst: Cache<Result<Arc<cst::decls::Module>, Error>>,
+    /// The symbol table constructed during lowering
+    cst_lookup_table: Cache<lowering::LookupTable>,
+    /// The AST of each file (once parsed and lowered, may be type-annotated)
+    ast: Cache<Result<Arc<ast::Module>, Error>>,
+    /// The symbol table constructed during typechecking
+    ast_lookup_table: Cache<elaborator::LookupTable>,
     info_by_id: HashMap<Url, Lapper<u32, Info>>,
```

Comment:
Should `info_by_id` and `item_by_id` not also be treated as a `Cache`? They can be invalidated, but we also want to provide e.g. type on hover using `get_even_if_stale` in the case that deps have changed.